### DISCUSSION TITLE
Colour aberration occurs when a PDF doc was converted to webp image

### DISCUSSION THREAD

**Discussion by scpzjj** [ImageMagick/ImageMagick#3294] [Help]
I tried to convert a PDF doc to Webp image, but there is a colour aberration between the origional PDF and the webp image. It seems the PDF doc is converted to a Greenish webp image. But if I convert the same PDF doc to jpeg image, the problem won't happen.

The origional PDF
![image](https://user-images.githubusercontent.com/16235140/109092815-38dd8f80-7752-11eb-95e5-1ccf628502a3.png)

Converted webp image
![image](https://user-images.githubusercontent.com/16235140/109092872-57438b00-7752-11eb-8879-3b9cf137900a.png)

Demo code I used is listed below.

class ImageMagick{
    private $handler;
    public function __construct()
    {
        $this->handler = new \Imagick();
    }
    public function convertPdf2Images($pdf_path,$out_path,$from=1,$to=1)
    {
        $this->handler->setResolution(120,120);
        $this->handler->setCompressionQuality(100);
        $this->handler->readImage($pdf_path);
        $pages = count($this->handler);
        if($pages == 0)
            throw new SystemInnerException('empty pdf file!');
        $index = 1;

        $data = [];
        foreach ($this->handler as $k=>$v)
        {
            if($index == $from)
            {
                if($from<=$to) {

                    $v->transformimagecolorspace(\Imagick::COLORSPACE_SRGB);
                    $v->setImageFormat('webp');
                    $tmp_name = strval($index) . '.webp';
                    if ($v->writeImage($out_path.$tmp_name) == true)
                        $data[] = $tmp_name;
                    $index++;
                    $from++;
                }else
                    break;
            }else
                $index++;
        }

        return $data;
    }
}
$imagick=new ImageMagick();
$imagick->convertPdf2Images("/usr/share/nginx/html/narrator/1.pdf", "/usr/share/nginx/html/narrator/", 1, 1);

**Comment by fmw42** [comment]
Your PDF file likely contains a raster image. That image may be CMYK or it may have a profile. If it is CMYK or has a profile, those may not be compatible with webp. Check your PDF or extract the image using pdf2image to see what colorspace or profile it has. Check the webp file to see what colorspace or profile it has.  Post your PDF file so that anyone can check it or test your processing.

**Reply by scpzjj** [reply]
You're right, the colorspace of the PDF doc is CMYK and the one of the converted webp is SRGB. Is there any way to reduce such color lost during conversion?

**Comment by snibgo** [comment]
@scpzjj: does the image witin the PDF have an embedded ICC profile? If it does, then you should convert it to an sRGB profile instead of using transformimagecolorspace.

EDIT: please link to your PDF.

**Reply by scpzjj** [reply]
I checked the PDF with 'identify' but no icc-profile was found. Does it mean color difference is inevitable when converting a PDF without icc profile to a Webp image?

**Comment by snibgo** [comment]
You show "The origional PDF" but that is a PNG sRGB image. What software converted the PDF to the PNG? How do you know that is the correct colour?

When there is no embedded ICC profile, we don't know the correct colours. We might assign a suitable ICC profile, then convert to a sRGB profile.

When we want to extract a raster image that is embedded in a PDF, I generally suggest using pdfimages, not ImageMagick.
